File: shared/relation_registry.go
package shared

import (
	_ "embed"
	"encoding/json"
	"strings"
)

//go:embed relation_types.json
var relationTypesJSON []byte

// RelationDefinition describes one semantic edge type shared by the frontend
// and backend.
type RelationDefinition struct {
	Code                string `json:"code"`
	Label               string `json:"label"`
	IsHierarchical      bool   `json:"hierarchical"`
	IsAcyclic           bool   `json:"acyclic"`
	AllowsMultiParent   bool   `json:"allows_multi_parent"`
	CanCollapseChildren bool   `json:"can_collapse_children"`
}

var relationDefinitions = mustLoadRelationDefinitions()

func mustLoadRelationDefinitions() map[string]RelationDefinition {
	var definitions []RelationDefinition
	if err := json.Unmarshal(relationTypesJSON, &definitions); err != nil {
		panic(err)
	}

	result := make(map[string]RelationDefinition, len(definitions))
	for _, definition := range definitions {
		code := strings.ToUpper(strings.TrimSpace(definition.Code))
		if code == "" {
			continue
		}

		definition.Code = code
		result[code] = definition
	}

	return result
}

// LookupRelationDefinition returns the canonical definition for one relation code.
func LookupRelationDefinition(code string) (RelationDefinition, bool) {
	definition, ok := relationDefinitions[strings.ToUpper(strings.TrimSpace(code))]
	return definition, ok
}

File: frontend/src/lib/relationRegistry.ts
import relationDefinitions from '../../../shared/relation_types.json';

export interface RelationDefinition {
  code: string;
  label: string;
  hierarchical: boolean;
  acyclic: boolean;
  allows_multi_parent: boolean;
  can_collapse_children: boolean;
}

export const DEFAULT_HIERARCHY_RELATION_TYPE = 'PARENT_CHILD';

const registry = new Map<string, RelationDefinition>(
  (relationDefinitions as RelationDefinition[]).map((definition) => [definition.code.trim().toUpperCase(), definition]),
);

export function getRelationDefinition(relationType: string): RelationDefinition | null {
  return registry.get(relationType.trim().toUpperCase()) ?? null;
}

export function isHierarchicalRelationType(relationType: string): boolean {
  return getRelationDefinition(relationType)?.hierarchical === true;
}

export function canCollapseChildrenForRelation(relationType: string): boolean {
  return getRelationDefinition(relationType)?.can_collapse_children === true;
}
